src/main/java/org/codehaus/mojo/flatten/FlattenMojo.java

```diff
@@ -657,18 +657,39 @@ protected Model createCleanPom( Model effectivePom ) throws MojoExecutionExcepti
         }
 
         // transform profiles...
+        Dependencies managedDependencies = new Dependencies();
+        if (effectivePom.getDependencyManagement() != null
+                && effectivePom.getDependencyManagement().getDependencies() != null)
+            managedDependencies.addAll(effectivePom.getDependencyManagement().getDependencies());
+
         for ( Profile profile : effectivePom.getProfiles() )
         {
             if ( !isEmbedBuildProfileDependencies() || !isBuildTimeDriven( profile.getActivation() ) )
             {
                 if ( !isEmpty( profile.getDependencies() ) || !isEmpty( profile.getRepositories() ) )
                 {
-                    Profile strippedProfile = new Profile();
-                    strippedProfile.setId( profile.getId() );
-                    strippedProfile.setActivation( profile.getActivation() );
-                    strippedProfile.setDependencies( profile.getDependencies() );
-                    strippedProfile.setRepositories( profile.getRepositories() );
-                    cleanPom.addProfile( strippedProfile );
+                    List<Dependency> strippedDependencies = new ArrayList<>();
+                    for (Dependency dep : profile.getDependencies()) {
+                        Dependency parsedDep = dep.clone();
+                        if (managedDependencies.contains(parsedDep)) {
+                            parsedDep.setVersion(managedDependencies.resolve(parsedDep).getVersion());
```

Comment:
last question - how do we know whether version needs to be set or not? i.e., set it if the final pom has no depMgmt section.

Reply:
Actually original flatten-maven-plugin will not decide whether version needs to be set or not - it just sets the version except we use "resolveCiFfriendliesOnly" mode.

When we set the <pomElement> and want to keep part of the element in the flattened pom, the flatten-maven-plugin will actually take it from effective pom. So if we want to keep the depMgmt in a flattened pom, it will be the combination of the depMgmt in original pom, the depMgmt inherited from parent pom, and the depMgmt imported. The version of dependencies will still be set whether we keep the depMgmt or not.